source/common/event/scaled_range_timer_manager_impl.cc

```diff
@@ -123,60 +123,105 @@ class ScaledRangeTimerManager::RangeTimerImpl final : public RangeTimer {
     }
   }
 
-  ScaledRangeTimerManager& manager_;
+  ScaledRangeTimerManagerImpl& manager_;
   const TimerCb callback_;
   const TimerPtr min_duration_timer_;
 
   absl::variant<Inactive, WaitingForMin, ScalingMax> state_;
   const ScopeTrackedObject* scope_;
 };
 
-ScaledRangeTimerManager::ScaledRangeTimerManager(Dispatcher& dispatcher)
+class ScaledRangeTimerManagerImpl::FixedMinimumRangeTimerImpl final : public Event::Timer {
+public:
+  FixedMinimumRangeTimerImpl(ScaledTimerMinimum minimum, TimerCb timer_cb,
+                             ScaledRangeTimerManagerImpl& manager)
+      : minimum_(minimum), range_timer_(std::move(timer_cb), manager) {}
+
+  void enableTimer(std::chrono::milliseconds ms,
+                   const ScopeTrackedObject* object = nullptr) override {
+    range_timer_.enableTimer(computeMinimum(minimum_, ms), ms, object);
+  }
+
+  void enableHRTimer(std::chrono::microseconds us,
+                     const ScopeTrackedObject* object = nullptr) override {
+    enableTimer(std::chrono::duration_cast<std::chrono::milliseconds>(us), object);
+  }
+
+  void disableTimer() override { range_timer_.disableTimer(); }
+
+  bool enabled() override { return range_timer_.enabled(); }
+
+private:
+  static std::chrono::milliseconds computeMinimum(const ScaledTimerMinimum minimum,
+                                                  std::chrono::milliseconds maximum) {
+    struct Visitor {
+      explicit Visitor(std::chrono::milliseconds value) : value_(value) {}
+      std::chrono::milliseconds operator()(ScaleFactor scale_factor) {
+        return std::chrono::duration_cast<std::chrono::milliseconds>(scale_factor.scale_factor_ *
+                                                                     value_);
+      }
+      std::chrono::milliseconds operator()(AbsoluteValue absolute_value) {
+        return absolute_value.value_;
+      }
+      const std::chrono::milliseconds value_;
+    };
+    return absl::visit(Visitor(maximum), minimum);
```

Comment:
Should computeMinimum be a method of ScaledTimerMinimum instead?  Or be a public helper method provided by scaled_range_timer_manager.h?  This visitor implementation is fairly non-trivial and there's no other practical way to interact with the absl::variant type used to hold the minimum, it would be best to avoid having multiple copies through the codebase.